### PR TITLE
feat(tests): new e2e tests [1/N]

### LINKED ISSUE #1650: [TRACKING] End-to-End Test Contribution Proposal
**Motivation**

As consumers of `libs`, we have developed a comprehensive set of end-to-end tests that have proven valuable in ensuring the reliability and functionality of `falcosecurity/libs`, particularly in conjunction with various drivers. Recognizing the benefits of community collaboration, we aim to upstream these tests to the `falcosecurity/libs` repository.

**Feature**

We propose the contribution of a set of private end-to-end tests specifically designed to validate the behaviour of `libsinsp` together with the associated drivers. These tests aim to cover critical scenarios, ensuring that the library performs optimally in various environments.

**Additional Context**

The proposed end-to-end tests cover a spectrum of scenarios, including but not limited to:

- Interaction of `libsinsp` with various drivers.
- Comprehensive coverage of edge cases and error scenarios.

These tests cover various areas of `libsinsp` interest, like:
- [x] processes
- [x] syscalls
- [x] cgroups
- [x] container runtimes
- [x] networking

To facilitate this process, we are creating this tracking issue to serve as a central point for discussion and coordination with the maintainers of `falcosecurity/libs`.

### PR DESCRIPTION
**Pull request by therealbobo**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

/area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap

> /area libpman

> /area libsinsp

/area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

This is the initial batch of new e2e tests.

**Which issue(s) this PR fixes**:

Part of #1650

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by leogr — COMMENTED** [review]
Hey @therealbobo 

Can you rebase this on the `master`?

This https://github.com/falcosecurity/libs/pull/1664 should fix the failing tests. :pray:

**Review by leogr — APPROVED** [review]
:partying_face:

File: test/libsinsp/sys_call_test.cpp [hunk 1]
```diff
@@ -0,0 +1,1235 @@
...
+#include <sys/types.h>
+
+#include <algorithm>
+#include <cassert>
+#include <list>
+#include <numeric>
+
+using namespace std;
+
+TEST_F(sys_call_test, stat)
```
Comment (Andreagit97) [inline test/libsinsp/sys_call_test.cpp:57]: Uhm, all the tests in this file seem to do the exact thing as `drivers_test`.
They run a syscall or a combination of syscalls and then assert that the event is here and something is happening in userspace... why don't we extend the `class event_test` to support some inspections in sinsp instead of introducing all these executables/classes + callbacks, WDYT?

**Comment by Andreagit97** [comment]
I left some initial comments but my main point is this one https://github.com/falcosecurity/libs/pull/1661#discussion_r1479660482

**Comment by LucaGuerra** [comment]
When attempting to build the test suite I get
```
In file included from /usr/include/features.h:392,
                 from /usr/include/arpa/inet.h:21,
                 from /home/ubuntu/dev/falcosecurity/libs/test/libsinsp/test_helper.cpp:1:
/usr/include/features-time64.h:20:10: fatal error: bits/wordsize.h: No such file or directory
   20 | #include <bits/wordsize.h>
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
```

Am I missing some dependency?

**Comment by therealbobo** [comment]
This is probably related to the build of the is needed gcc with multilib support: there are some tests that run also for x86.

File: .github/workflows/e2e_ci.yml [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+      - name: Run e2e tests with bpf 🏎️
+        run: |
+          cd build/test/libsinsp/
+          sudo ./tests -b
+
+      - name: Run e2e tests with kmod 🏎️
+        run: |
+          cd build/test/libsinsp/
+          sudo ./tests -k
```
[the remark this answers is not in the document]
Comment (jasondellaluce) [reply .github/workflows/e2e_ci.yml:126]: Another suggestion I'd add here is that we may want to run the tests on different drivers by using a matrix without soft fail, so that we don't skip tests on a driver if we find errors in another.

File: test/libsinsp/libsinsp_test_var.h.in [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+// Absolute path to the kernel module file
+#define LIBSINSP_TEST_KERNEL_MODULE_PATH "${CMAKE_BINARY_DIR}/driver/scap.ko"
+
+// Kernel module name
+#define LIBSINSP_TEST_KERNEL_MODULE_NAME "${DRIVER_NAME}"
+
+// Absolute path to the bpf probe .o file
+#define LIBSINSP_TEST_BPF_PROBE_PATH "${CMAKE_BINARY_DIR}/driver/bpf/probe.o"
+
+#define LIBSINSP_TEST_CAPTURES_PATH "${CMAKE_BINARY_DIR}/test/libsinsp/captures"
```
[the remark this answers is not in the document]
Comment (therealbobo) [reply test/libsinsp/libsinsp_test_var.h.in:30]: `LIBSINSP_TEST_CAPTURES_PATH` has a different motivation: each test dumps the captured events in a scap file so that we can inspect them in a second moment. This is not meant to test scap file correctness. 😅

File: test/libsinsp/sys_call_test.cpp (hunk 1 above)
Comment (jasondellaluce) [reply test/libsinsp/sys_call_test.cpp:57]: Although some tests are similar in this first batch, the whole contribution process of the test suite we foresee can potentially cover use cases that diverge with the goals of the driver-level tests. It's hard to predict it at the moment, and refactor everything right now may be outside of the engineering bandwidth we currently have available. Plus, having this suite being its own standalone thing will allow us to converge it with the drivers one once we have a more high-level view than now. If this was a library code contribution, I'd agree with you of this being a strong blocker. But since this is a test-based contribution, I think it can provide tons of value as-is and we can postpone cleaning it up deeply as a secondary priority. Moreover @therealbobo is also curating each test while contributing them.

File: test/libsinsp/CMakeLists.txt
[the remark this answers is not in the document]
Comment (therealbobo) [reply test/libsinsp/CMakeLists.txt:1]: Yes, totally on board on this! On the other hand I'd like to do this kind on changes later on (I already have some PRs in the working). Are you ok with that?

File: .github/workflows/e2e_ci.yml [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+      - name: Install kernel headers and gcc
+        if: matrix.arch == 'amd64'
+        run: |
+          sudo apt install -y --no-install-recommends linux-headers-$(uname -r) gcc-multilib g++-multilib
+
+      - name: Build e2e tests 🏗️
+        run: |
+          mkdir -p build
+          cd build && cmake -DUSE_BUNDLED_DEPS=ON -DENABLE_LIBSINSP_TESTS=ON -DBUILD_LIBSCAP_MODERN_BPF=ON -DBUILD_BPF=ON -DBUILD_LIBSCAP_GVISOR=OFF ../
```
Comment (LucaGuerra) [inline .github/workflows/e2e_ci.yml:62]: It looks like `ENABLE_LIBSINSP_TESTS` was renamed to `ENABLE_LIBSINSP_E2E_TESTS`

File: test/libsinsp_e2e/sys_call_test.cpp [hunk 5]
```diff
@@ -505,7 +506,8 @@ TEST_F(sys_call_test, brk)
 
 		if (type == PPME_SYSCALL_BRK_4_E)
 		{
-			uint64_t addr = *((uint64_t*)e->get_param_by_name("addr")->m_val);
+			uint64_t addr;
+			memcpy(&addr,e->get_param_by_name("addr")->m_val, sizeof(uint64_t));
```
Comment (LucaGuerra) [inline test/libsinsp_e2e/sys_call_test.cpp:510]: Thanks for these fixese, but it's even easier than this!

```c++
e->get_param_by_name("addr")->as<uint64_t>();
```

😎

Comment (therealbobo) [reply test/libsinsp_e2e/sys_call_test.cpp:510]: I totally forgot about that! Thanks!

**Comment by LucaGuerra** [comment]
A couple more issues I have encountered while trying this out. While the 64 bit tests work (great job! 🎉 ) the 32 bit have issues:
* If I run the tests like this: `sudo ./test/libsinsp_e2e/libsinsp_e2e_tests  -m --gtest_filter=sys_call_test32.*` (as described in the README) I get 
```
[ RUN      ] sys_call_test32.execve_ia32_emulation
sh: 1: ./resources/execve32: not found
/home/ubuntu/dev/falcosecurity/libs/test/libsinsp_e2e/sys_call_test.cpp:1010: Failure
```
Probably just a matter of setting the right directory
* In addition, I have built the tests by running `make -j10 libsinsp_e2e_tests` but even if I run the tests from their `./test/libsinsp_e2e` directory I see an attempt to run `./test_helper_32` which is not found:
```
[ RUN      ] sys_call_test32.quotactl_ko
Running: ./test_helper_32
Failed to execute the process: No such file or directory
```

**Comment by LucaGuerra** [comment]
Also I can confirm that these tests (except for the two cases above) are passing with ASan and UBSan enabled ❤️ , thanks to @therealbobo for working with me to fix the warnings

File: test/libsinsp_e2e/subprocess.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+        std::vector<char*> args;
+        args.push_back(const_cast<char*>(command.c_str()));
+        for (const auto& arg : arguments) {
+            args.push_back(const_cast<char*>(arg.c_str()));
+        }
+        args.push_back(nullptr);
+
+        execvp(command.c_str(), args.data());
+        std::cerr << "Failed to execute the process." << std::endl;
```
Comment (LucaGuerra) [inline test/libsinsp_e2e/subprocess.cpp:65]:
```suggestion
        std::cerr << "Failed to execute " << command << ": " << std::strerror(errno) << std::endl;
```
I made this modification to have a better error

Comment (LucaGuerra) [reply test/libsinsp_e2e/subprocess.cpp:65]: (you may also need to include errno.h)

File: test/libsinsp_e2e/sys_call_test.cpp [hunk 7]
```diff
@@ -0,0 +1,1237 @@
...
+
+	event_filter_t filter = [&](sinsp_evt* evt)
+	{
+		return evt->get_type() == PPME_SYSCALL_QUOTACTL_X ||
+		       evt->get_type() == PPME_SYSCALL_QUOTACTL_E;
+	};
+
+	run_callback_t test = [&](sinsp* inspector)
+	{
+		subprocess handle("./test_helper_32", {"quotactl_ko"});
```
Comment (mrgian) [inline test/libsinsp_e2e/sys_call_test.cpp:1077]:
```suggestion
		subprocess handle("./test/libsinsp_e2e/test_helper_32", {"quotactl_ko"});
```

**Review by mrgian — COMMENTED** [review]
> Probably just a matter of setting the right directory

Hi @LucaGuerra 
You are right, it's just a path issue caused by the new name of the test's directory.
I've tried running these tests with the correct paths and they are working 🎉

File: test/libsinsp_e2e/sys_call_test.cpp [hunk 8]
```diff
@@ -0,0 +1,1237 @@
...
+		sinsp_filter_compiler compiler(inspector,
+		                               "evt.type=execve and proc.apid=" + std::to_string(getpid()));
+		is_subprocess_execve.reset(compiler.compile());
+	};
+
+	event_filter_t filter = [&](sinsp_evt* evt) { return is_subprocess_execve->run(evt); };
+
+	run_callback_t test = [&](sinsp* inspector)
+	{
+		auto ret = system("./resources/execve32 ./resources/execve ./resources/execve32");
```
[the remark this answers is not in the document]
Comment (therealbobo) [reply test/libsinsp_e2e/sys_call_test.cpp]: I forgot to push the fix 😅 This commit should fix it once and for all: https://github.com/falcosecurity/libs/pull/1661/commits/0009a014d6e3b6dd5f80fbf037bfaecb5dd1fe62. It sets the absolute paths a priori.

[the remark this answers is not in the document]
Comment (leogr) [reply test/libsinsp_e2e/sys_call_test.cpp]: cc @LucaGuerra

**Comment by LucaGuerra** [comment]
I had tested this before and it seems like cmake had issues generating the binary, I'll clean and retry. Thank you for your efforts in making the e2e tests more compatible!

**Review by LucaGuerra — APPROVED** [review]
After the latest fixes I was able to run the tests as specified in the readme. Thanks!

**Review by leogr — APPROVED** [review]

**Comment by LucaGuerra** [comment]
/unhold

I believe review comments have been addressed
